File: scripts/phase1-locator-batch-a.mjs
#!/usr/bin/env node

import { readFile, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { getDataset } from '../src/data/sovereigns.merge.js';
import { sourceById } from '../src/data/sources.js';

const DATE = '2026-02-08';
const OWNER = 'phase1-locator-a';
const __dirname = dirname(fileURLToPath(import.meta.url));
const rootDir = join(__dirname, '..');
const ledgerDir = join(rootDir, 'docs', 'research-program', 'ledgers');

const claimUpdates = new Map([
  ['CLM-0236', {
    claim_excerpt: 'Official former-president profile states Mohamed Amin was born as son of Ahmed Didi and Nayaage Aishath Didi.',
    citation_locator: 'SRC-PO-AMIN profile (presidency.gov.mv/PO/FormerPresident/7), lines 49-52 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from official profile text.'
  }],
  ['CLM-0271', {
    claim_excerpt: 'Official former-president profile states Mohamed Amin was born as son of Ahmed Didi and Nayaage Aishath Didi.',
    citation_locator: 'SRC-PO-AMIN profile (presidency.gov.mv/PO/FormerPresident/7), lines 49-52 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from official profile text.'
  }],
  ['CLM-0244', {
    claim_excerpt: 'Official former-president profile states Ibrahim Nasir was son of Ahmed Didi and Nayaage Aishath Didi.',
    citation_locator: 'SRC-PO-NASIR profile (presidency.gov.mv/PO/FormerPresident/6), lines 48-50 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from official profile text.'
  }],
  ['CLM-0245', {
    claim_excerpt: 'Official former-president profile states Ibrahim Nasir was son of Ahmed Didi and Nayaage Aishath Didi.',
    citation_locator: 'SRC-PO-NASIR profile (presidency.gov.mv/PO/FormerPresident/6), lines 48-50 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from official profile text.'
  }],
  ['CLM-0253', {
    claim_excerpt: 'Official former-president profile states Abdulla Yameen is son of Abdul Gayoom Ibrahim.',
    citation_locator: 'SRC-PO-YAMEEN profile (presidency.gov.mv/PO/FormerPresident/1), lines 43-44 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from official profile text.'
  }],
  ['CLM-0375', {
    claim_excerpt: 'Official former-president profile identifies Abdulla Yameen as younger brother of former President Maumoon Abdul Gayoom.',
    citation_locator: 'SRC-PO-YAMEEN profile (presidency.gov.mv/PO/FormerPresident/1), lines 44-45 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from official profile text.'
  }],
  ['CLM-0258', {
    claim_excerpt: 'Atoll Times report names Mohamed Nasheed’s father as Abdul Sattar Umar.',
    citation_locator: 'SRC-ATOLL-NASHEED-PARENTS article (atolltimes.mv/post/news/3844), lines 27-28 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from named parent statement in news report.'
  }],
  ['CLM-0259', {
    claim_excerpt: 'Atoll Times report names Mohamed Nasheed’s mother as Abida Mohamed.',
    citation_locator: 'SRC-ATOLL-NASHEED-PARENTS article (atolltimes.mv/post/news/3844), lines 27-28 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from named parent statement in news report.'
  }],
  ['CLM-0307', {
    claim_excerpt: 'Edition obituary report identifies Aishath Khadheeja as mother of former President Ibrahim Mohamed Solih.',
    citation_locator: 'SRC-EDITION-SOLIH-MOTHER report (edition.mv/ibrahim_muaz_mua/34704), line 31 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from obituary identification text.'
  }],
  ['CLM-0297', {
    claim_excerpt: 'Wikipedia personal-life section lists Widhadh Waheed among Dr Mohamed Waheed Hassan’s children.',
    citation_locator: 'SRC-WIKI-WAHEED page (en.wikipedia.org/wiki/Mohamed_Waheed_Hassan), lines 56-57 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from personal-life child listing.'
  }],
  ['CLM-0298', {
    claim_excerpt: 'Wikipedia personal-life section lists Fidha Waheed among Dr Mohamed Waheed Hassan’s children.',
    citation_locator: 'SRC-WIKI-WAHEED page (en.wikipedia.org/wiki/Mohamed_Waheed_Hassan), lines 56-57 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from personal-life child listing.'
  }],
  ['CLM-0299', {
    claim_excerpt: 'Wikipedia personal-life section lists Jeffrey Salim Waheed among Dr Mohamed Waheed Hassan’s children.',
    citation_locator: 'SRC-WIKI-WAHEED page (en.wikipedia.org/wiki/Mohamed_Waheed_Hassan), lines 56-57 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from personal-life child listing.'
  }],
  ['CLM-0300', {
    claim_excerpt: 'Wikipedia personal-life section lists Widhadh Waheed among Dr Mohamed Waheed Hassan and Ilham Hussain’s children.',
    citation_locator: 'SRC-WIKI-WAHEED page (en.wikipedia.org/wiki/Mohamed_Waheed_Hassan), lines 48 and 56-57 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from spouse and child listing.'
  }],
  ['CLM-0301', {
    claim_excerpt: 'Wikipedia personal-life section lists Fidha Waheed among Dr Mohamed Waheed Hassan and Ilham Hussain’s children.',
    citation_locator: 'SRC-WIKI-WAHEED page (en.wikipedia.org/wiki/Mohamed_Waheed_Hassan), lines 48 and 56-57 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from spouse and child listing.'
  }],
  ['CLM-0302', {
    claim_excerpt: 'Wikipedia personal-life section lists Jeffrey Salim Waheed among Dr Mohamed Waheed Hassan and Ilham Hussain’s children.',
    citation_locator: 'SRC-WIKI-WAHEED page (en.wikipedia.org/wiki/Mohamed_Waheed_Hassan), lines 48 and 56-57 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from spouse and child listing.'
  }],
  ['CLM-0049', {
    claim_excerpt: 'Wikipedia lead text identifies Ibrahim Mohamed Solih as first cousin of former President Mohamed Nasheed.',
    citation_locator: 'SRC-WIKI-SOLIH page (en.wikipedia.org/wiki/Ibrahim_Mohamed_Solih), line 34 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead relationship statement.'
  }],
  ['CLM-0304', {
    claim_excerpt: 'Wikipedia lead text states Ibrahim Mohamed Solih and his wife have a son named Yaman.',
    citation_locator: 'SRC-WIKI-SOLIH page (en.wikipedia.org/wiki/Ibrahim_Mohamed_Solih), lines 43-45 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead family statement.'
  }],
  ['CLM-0306', {
    claim_excerpt: 'Wikipedia lead text states Ibrahim Mohamed Solih and his wife have a son named Yaman.',
    citation_locator: 'SRC-WIKI-SOLIH page (en.wikipedia.org/wiki/Ibrahim_Mohamed_Solih), lines 43-45 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead family statement.'
  }],
  ['CLM-0460', {
    claim_excerpt: 'Wikipedia lead text names Fazeena Ahmed as spouse of Ibrahim Mohamed Solih.',
    citation_locator: 'SRC-WIKI-SOLIH page (en.wikipedia.org/wiki/Ibrahim_Mohamed_Solih), lines 43-45 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead family statement.'
  }],
  ['CLM-0314', {
    claim_excerpt: 'Wikipedia lead text states Mohamed Muizzu was born to Husna Adam Ismail and Hussain Abdul Rahman.',
    citation_locator: 'SRC-WIKI-MUIZZU page (en.wikipedia.org/wiki/Mohamed_Muizzu), lines 53-54 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead parentage statement.'
  }],
  ['CLM-0315', {
    claim_excerpt: 'Wikipedia lead text states Mohamed Muizzu was born to Husna Adam Ismail and Hussain Abdul Rahman.',
    citation_locator: 'SRC-WIKI-MUIZZU page (en.wikipedia.org/wiki/Mohamed_Muizzu), lines 53-54 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead parentage statement.'
  }],
  ['CLM-0461', {
    claim_excerpt: 'Wikipedia lead text names Laila (Sajidha Mohamed) as spouse of Mohamed Muizzu.',
    citation_locator: 'SRC-WIKI-MUIZZU page (en.wikipedia.org/wiki/Mohamed_Muizzu), line 56 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead spouse statement.'
  }],
  ['CLM-0457', {
    claim_excerpt: 'Wikipedia lead text states Mohamed Nasheed is married to Laila Ali Abdulla.',
    citation_locator: 'SRC-WIKI-NASHEED page (en.wikipedia.org/wiki/Mohamed_Nasheed), line 50 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead spouse statement.'
  }],
  ['CLM-0456', {
    claim_excerpt: 'Wikipedia lead text states Maumoon Abdul Gayoom is married to Nasreena Ibrahim.',
    citation_locator: 'SRC-WIKI-MAUMOON page (en.wikipedia.org/wiki/Maumoon_Abdul_Gayoom), line 29 in 2026-02-08 extraction snapshot.',
    notes: 'Locator extraction completed from lead spouse statement.'
  }]
]);

const sourceExtractEntries = [
  {
    extract_id: 'EXT-016',
    source_id: 'SRC-PO-AMIN',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Official parentage statement extraction',
    claim_or_context_note: 'Extracted parentage phrasing for Mohamed Amin profile and mapped to parent claims.',
    locator: 'presidency FormerPresident/7 lines 49-52 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0236 and CLM-0271.'
  },
  {
    extract_id: 'EXT-017',
    source_id: 'SRC-PO-NASIR',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Official parentage statement extraction',
    claim_or_context_note: 'Extracted parentage phrasing for Ibrahim Nasir profile and mapped to parent claims.',
    locator: 'presidency FormerPresident/6 lines 48-50 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0244 and CLM-0245.'
  },
  {
    extract_id: 'EXT-018',
    source_id: 'SRC-PO-YAMEEN',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Official parent/sibling statement extraction',
    claim_or_context_note: 'Extracted father and sibling wording for Yameen profile.',
    locator: 'presidency FormerPresident/1 lines 43-45 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0253 and CLM-0375.'
  },
  {
    extract_id: 'EXT-019',
    source_id: 'SRC-ATOLL-NASHEED-PARENTS',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Named parent statement extraction',
    claim_or_context_note: 'Extracted explicit naming of Nasheed father and mother from report text.',
    locator: 'atolltimes article lines 27-28 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0258 and CLM-0259.'
  },
  {
    extract_id: 'EXT-020',
    source_id: 'SRC-EDITION-SOLIH-MOTHER',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Maternal identity statement extraction',
    claim_or_context_note: 'Extracted obituary phrasing identifying former President Solih’s mother.',
    locator: 'edition report line 31 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0307.'
  },
  {
    extract_id: 'EXT-021',
    source_id: 'SRC-WIKI-WAHEED',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Personal-life family listing extraction',
    claim_or_context_note: 'Extracted spouse and children listing for Waheed family claims.',
    locator: 'Wikipedia lines 48,56-57 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0297 through CLM-0302.'
  },
  {
    extract_id: 'EXT-022',
    source_id: 'SRC-WIKI-SOLIH',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Lead family relation extraction',
    claim_or_context_note: 'Extracted first-cousin and spouse/son statements for Solih.',
    locator: 'Wikipedia lines 34,43-45 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0049, CLM-0304, CLM-0306, CLM-0460.'
  },
  {
    extract_id: 'EXT-023',
    source_id: 'SRC-WIKI-MUIZZU',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Lead parentage and spouse extraction',
    claim_or_context_note: 'Extracted parent names and spouse naming from Muizzu lead text.',
    locator: 'Wikipedia lines 53-54 and 56 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0314, CLM-0315, CLM-0461.'
  },
  {
    extract_id: 'EXT-024',
    source_id: 'SRC-WIKI-NASHEED',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Lead spouse extraction',
    claim_or_context_note: 'Extracted spouse naming from Nasheed lead text.',
    locator: 'Wikipedia line 50 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0457.'
  },
  {
    extract_id: 'EXT-025',
    source_id: 'SRC-WIKI-MAUMOON',
    date: DATE,
    extract_status: 'in_progress',
    target_scope: 'Lead spouse extraction',
    claim_or_context_note: 'Extracted spouse naming from Maumoon lead text.',
    locator: 'Wikipedia line 29 snapshot',
    researcher: OWNER,
    notes: 'Supports CLM-0456.'
  }
];

const queueEntries = [
  {
    queue_id: 'SRCQ-016',
    priority: 'P1',
    track: 'Official profiles',
    dynasty_or_topic: 'Modern political lineage anchors',
    target_claim_scope: 'Parent and sibling claims from official presidential profiles',
    candidate_source_id: 'SRC-PO-AMIN|SRC-PO-NASIR|SRC-PO-YAMEEN',
    source_url: '',
    publisher: "President's Office Maldives",
    source_type: 'official profile',
    expected_grade: 'A',
    status: 'in_progress',
    last_updated: DATE,
    owner: OWNER,
    notes: 'Locator extraction batch A completed for selected parent/sibling statements (EXT-016..018).'
  },
  {
    queue_id: 'SRCQ-017',
    priority: 'P1',
    track: 'Modern family corroboration',
    dynasty_or_topic: 'Modern nodes P115-P173 cluster',
    target_claim_scope: 'Spouse, parent, and cousin statements with explicit lead/personal-life locators',
    candidate_source_id: 'SRC-WIKI-WAHEED|SRC-WIKI-SOLIH|SRC-WIKI-MUIZZU|SRC-WIKI-NASHEED|SRC-WIKI-MAUMOON',
    source_url: '',
    publisher: 'Wikipedia',
    source_type: 'secondary synthesis',
    expected_grade: 'B',
    status: 'in_progress',
    last_updated: DATE,
    owner: OWNER,
    notes: 'Locator extraction batch A completed for explicit lead/personal-life statements (EXT-021..025).'
  },
  {
    queue_id: 'SRCQ-018',
    priority: 'P1',
    track: 'News corroboration',
    dynasty_or_topic: 'Modern family parentage claims',
    target_claim_scope: 'Named parent statements in contemporary reports',
    candidate_source_id: 'SRC-ATOLL-NASHEED-PARENTS|SRC-EDITION-SOLIH-MOTHER',
    source_url: '',
    publisher: 'Atoll Times / The Edition',
    source_type: 'news report',
    expected_grade: 'B',
    status: 'in_progress',
    last_updated: DATE,
    owner: OWNER,
    notes: 'Locator extraction batch A completed for explicit named-parent statements (EXT-019..020).'
  }
];

function parseCsv(text) {
  const rows = [];
  let row = [];
  let field = '';
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i += 1;
        } else {
          inQuotes = false;
        }
      } else {
        field += ch;
      }
      continue;
    }
    if (ch === '"') {
      inQuotes = true;
      continue;
    }
    if (ch === ',') {
      row.push(field);
      field = '';
      continue;
    }
    if (ch === '\n') {
      row.push(field);
      rows.push(row);
      row = [];
      field = '';
      continue;
    }
    if (ch !== '\r') field += ch;
  }
  if (field.length || row.length) {
    row.push(field);
    rows.push(row);
  }
  return rows;
}

function csvEscape(v) {
  const s = String(v ?? '');
  if (s.includes('"') || s.includes(',') || s.includes('\n')) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}

function toCsv(rows) {
  return `${rows.map(r => r.map(csvEscape).join(',')).join('\n')}\n`;
}

function byIdMap(list) {
  return new Map(list.map(item => [item.id, item]));
}

function personLabel(byId, id) {
  const p = byId.get(id);
  if (!p) return id;
  return `${id} ${p.nm || '(unnamed)'}`;
}

function applyRelationshipUpdates(dataset) {
  const path = join(ledgerDir, 'relationship-evidence-ledger.csv');
  return readFile(path, 'utf8').then(text => {
    const rows = parseCsv(text);
    const header = rows[0];
    const idx = Object.fromEntries(header.map((h, i) => [h, i]));
    const byPersonId = byIdMap(dataset.people);
    let changed = 0;

    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      const claimId = row[idx.claim_id];
      const update = claimUpdates.get(claimId);
      if (!update) continue;

      row[idx.claim_excerpt] = update.claim_excerpt;
      row[idx.citation_locator] = update.citation_locator;
      row[idx.review_status] = 'in_progress';
      row[idx.canonical_decision] = 'pending';
      row[idx.reviewer] = OWNER;
      row[idx.last_reviewed] = DATE;
      row[idx.access_date] = DATE;
      row[idx.notes] = update.notes;

      const sid = row[idx.source_id];
      const tid = row[idx.target_id];
      row[idx.claim_excerpt] = `${update.claim_excerpt} (pair: ${personLabel(byPersonId, sid)} -> ${personLabel(byPersonId, tid)}).`;
      changed += 1;
    }
    return writeFile(path, toCsv(rows), 'utf8').then(() => changed);
  });
}

async function upsertSourceExtractLog() {
  const path = join(ledgerDir, 'source-extract-log.csv');
  const rows = parseCsv(await readFile(path, 'utf8'));
  const header = rows[0];
  const idx = Object.fromEntries(header.map((h, i) => [h, i]));

  const byId = new Map();
  for (let i = 1; i < rows.length; i++) {
    byId.set(rows[i][idx.extract_id], i);
  }

  for (const entry of sourceExtractEntries) {
    const row = header.map(col => entry[col] ?? '');
    if (byId.has(entry.extract_id)) {
      rows[byId.get(entry.extract_id)] = row;
    } else {
      rows.push(row);
    }
  }

  const body = rows.slice(1).sort((a, b) => a[idx.extract_id].localeCompare(b[idx.extract_id], undefined, { numeric: true }));
  await writeFile(path, toCsv([header, ...body]), 'utf8');
  return sourceExtractEntries.length;
}

async function upsertSourceQueue() {
  const path = join(ledgerDir, 'source-expansion-queue.csv');
  const rows = parseCsv(await readFile(path, 'utf8'));
  const header = rows[0];
  const idx = Object.fromEntries(header.map((h, i) => [h, i]));
  const byId = new Map();
  for (let i = 1; i < rows.length; i++) {
    byId.set(rows[i][idx.queue_id], i);
  }

  for (const entry of queueEntries) {
    const row = header.map(col => entry[col] ?? '');
    if (byId.has(entry.queue_id)) {
      rows[byId.get(entry.queue_id)] = row;
    } else {
      rows.push(row);
    }
  }

  const body = rows.slice(1).sort((a, b) => a[idx.queue_id].localeCompare(b[idx.queue_id], undefined, { numeric: true }));
  await writeFile(path, toCsv([header, ...body]), 'utf8');
  return queueEntries.length;
}

async function checkSources() {
  const missing = [];
  for (const entry of sourceExtractEntries) {
    if (!sourceById.get(entry.source_id)) missing.push(entry.source_id);
  }
  if (missing.length) {
    throw new Error(`Missing source IDs in registry: ${missing.join(', ')}`);
  }
}

async function main() {
  await checkSources();
  const dataset = getDataset('research');
  const claimCount = await applyRelationshipUpdates(dataset);
  const extractCount = await upsertSourceExtractLog();
  const queueCount = await upsertSourceQueue();
  console.log(`Phase 1 locator batch A complete:
- relationship claims updated: ${claimCount}
- source extract entries upserted: ${extractCount}
- source queue entries upserted: ${queueCount}`);
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: crates/research-cli/src/phase1_locator_batch_a.rs
use crate::csv_utils::{csv_escape, parse_csv};
use crate::source_models::SourceRecord;
use maldives_domain::Dataset;
use serde::Deserialize;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::Path;
use std::sync::OnceLock;

const OWNER: &str = "phase1-locator-a";
const DEFAULT_DATE: &str = "2026-02-08";

#[derive(Clone, Debug, Deserialize)]
struct ClaimUpdate {
    claim_excerpt: String,
    citation_locator: String,
    notes: String,
}

#[derive(Clone, Debug, Deserialize)]
struct Payload {
    claim_updates: HashMap<String, ClaimUpdate>,
    source_extract_entries: Vec<HashMap<String, String>>,
    queue_entries: Vec<HashMap<String, String>>,
}

fn payload() -> &'static Payload {
    static PAYLOAD: OnceLock<Payload> = OnceLock::new();
    PAYLOAD.get_or_init(|| {
        serde_json::from_str(include_str!("../data/phase1_locator_batch_a.payload.json"))
            .expect("valid phase1 locator batch A payload")
    })
}

fn rows_to_csv(rows: &[Vec<String>]) -> String {
    let mut out = String::new();
    for row in rows {
        let line = row
            .iter()
            .map(|v| csv_escape(v))
            .collect::<Vec<_>>()
            .join(",");
        out.push_str(&line);
        out.push('\n');
    }
    out
}

fn header_index(header: &[String]) -> HashMap<String, usize> {
    header
        .iter()
        .enumerate()
        .map(|(i, h)| (h.clone(), i))
        .collect::<HashMap<_, _>>()
}

fn row_get<'a>(row: &'a [String], idx: &HashMap<String, usize>, key: &str) -> &'a str {
    idx.get(key)
        .and_then(|i| row.get(*i))
        .map(String::as_str)
        .unwrap_or("")
}

fn row_set(row: &mut [String], idx: &HashMap<String, usize>, key: &str, value: String) {
    let Some(col) = idx.get(key).copied() else {
        return;
    };
    if col < row.len() {
        row[col] = value;
    }
}

fn by_id_map(dataset: &Dataset) -> HashMap<String, String> {
    dataset
        .people
        .iter()
        .map(|person| {
            let name = if person.nm.is_empty() {
                "(unnamed)".to_string()
            } else {
                person.nm.clone()
            };
            (person.id.clone(), format!("{} {}", person.id, name))
        })
        .collect::<HashMap<_, _>>()
}

fn person_label(by_id: &HashMap<String, String>, id: &str) -> String {
    by_id.get(id).cloned().unwrap_or_else(|| id.to_string())
}

fn parse_prefixed_numeric_id(value: &str) -> (&str, Option<u64>) {
    if let Some((prefix, number)) = value.split_once('-') {
        if let Ok(n) = number.parse::<u64>() {
            return (prefix, Some(n));
        }
    }
    (value, None)
}

fn compare_prefixed_numeric(a: &str, b: &str) -> std::cmp::Ordering {
    let (a_prefix, a_num) = parse_prefixed_numeric_id(a);
    let (b_prefix, b_num) = parse_prefixed_numeric_id(b);
    if a_prefix != b_prefix {
        return a.cmp(b);
    }
    match (a_num, b_num) {
        (Some(na), Some(nb)) => na.cmp(&nb),
        _ => a.cmp(b),
    }
}

fn with_date(value: &str, date: &str) -> String {
    if date == DEFAULT_DATE {
        value.to_string()
    } else {
        value.replace(DEFAULT_DATE, date)
    }
}

fn apply_relationship_updates(dataset: &Dataset, path: &Path, date: &str) -> Result<usize, String> {
    let rel_raw =
        fs::read_to_string(path).map_err(|e| format!("failed to read {}: {e}", path.display()))?;
    let mut rel_rows = parse_csv(&rel_raw);
    if rel_rows.is_empty() {
        return Err(format!("empty csv: {}", path.display()));
    }
    let rel_idx = header_index(&rel_rows[0]);
    let by_person_id = by_id_map(dataset);

    let mut changed = 0usize;
    for row in rel_rows.iter_mut().skip(1) {
        let claim_id = row_get(row, &rel_idx, "claim_id");
        let Some(update) = payload().claim_updates.get(claim_id) else {
            continue;
        };
        let source_id = row_get(row, &rel_idx, "source_id").to_string();
        let target_id = row_get(row, &rel_idx, "target_id").to_string();
        let source = person_label(&by_person_id, &source_id);
        let target = person_label(&by_person_id, &target_id);

        row_set(
            row,
            &rel_idx,
            "claim_excerpt",
            format!("{} (pair: {} -> {}).", update.claim_excerpt, source, target),
        );
        row_set(
            row,
            &rel_idx,
            "citation_locator",
            with_date(&update.citation_locator, date),
        );
        row_set(row, &rel_idx, "review_status", "in_progress".to_string());
        row_set(row, &rel_idx, "canonical_decision", "pending".to_string());
        row_set(row, &rel_idx, "reviewer", OWNER.to_string());
        row_set(row, &rel_idx, "last_reviewed", date.to_string());
        row_set(row, &rel_idx, "access_date", date.to_string());
        row_set(row, &rel_idx, "notes", update.notes.clone());
        changed += 1;
    }

    fs::write(path, rows_to_csv(&rel_rows))
        .map_err(|e| format!("failed to write {}: {e}", path.display()))?;
    Ok(changed)
}

fn upsert_source_extract_log(path: &Path, date: &str) -> Result<usize, String> {
    let ext_raw =
        fs::read_to_string(path).map_err(|e| format!("failed to read {}: {e}", path.display()))?;
    let mut ext_rows = parse_csv(&ext_raw);
    if ext_rows.is_empty() {
        return Err(format!("empty csv: {}", path.display()));
    }
    let ext_header = ext_rows[0].clone();
    let ext_idx = header_index(&ext_header);

    let mut by_id = HashMap::<String, usize>::new();
    for (i, row) in ext_rows.iter().enumerate().skip(1) {
        by_id.insert(row_get(row, &ext_idx, "extract_id").to_string(), i);
    }

    for entry in &payload().source_extract_entries {
        let extract_id = entry.get("extract_id").cloned().unwrap_or_default();
        let row_data = ext_header
            .iter()
            .map(|col| {
                with_date(
                    entry
                        .get(col.as_str())
                        .map(String::as_str)
                        .unwrap_or_default(),
                    date,
                )
            })
            .collect::<Vec<_>>();
        if let Some(i) = by_id.get(&extract_id).copied() {
            ext_rows[i] = row_data;
        } else {
            ext_rows.push(row_data);
        }
    }

    let mut ext_body = ext_rows.into_iter().skip(1).collect::<Vec<_>>();
    ext_body.sort_by(|a, b| {
        compare_prefixed_numeric(
            row_get(a, &ext_idx, "extract_id"),
            row_get(b, &ext_idx, "extract_id"),
        )
    });
    let mut ext_out = vec![ext_header];
    ext_out.extend(ext_body);
    fs::write(path, rows_to_csv(&ext_out))
        .map_err(|e| format!("failed to write {}: {e}", path.display()))?;

    Ok(payload().source_extract_entries.len())
}

fn upsert_source_queue(path: &Path, date: &str) -> Result<usize, String> {
    let queue_raw =
        fs::read_to_string(path).map_err(|e| format!("failed to read {}: {e}", path.display()))?;
    let mut queue_rows = parse_csv(&queue_raw);
    if queue_rows.is_empty() {
        return Err(format!("empty csv: {}", path.display()));
    }
    let queue_header = queue_rows[0].clone();
    let queue_idx = header_index(&queue_header);

    let mut by_id = HashMap::<String, usize>::new();
    for (i, row) in queue_rows.iter().enumerate().skip(1) {
        by_id.insert(row_get(row, &queue_idx, "queue_id").to_string(), i);
    }

    for entry in &payload().queue_entries {
        let queue_id = entry.get("queue_id").cloned().unwrap_or_default();
        let row_data = queue_header
            .iter()
            .map(|col| {
                with_date(
                    entry
                        .get(col.as_str())
                        .map(String::as_str)
                        .unwrap_or_default(),
                    date,
                )
            })
            .collect::<Vec<_>>();
        if let Some(i) = by_id.get(&queue_id).copied() {
            queue_rows[i] = row_data;
        } else {
            queue_rows.push(row_data);
        }
    }

    let mut queue_body = queue_rows.into_iter().skip(1).collect::<Vec<_>>();
    queue_body.sort_by(|a, b| {
        compare_prefixed_numeric(
            row_get(a, &queue_idx, "queue_id"),
            row_get(b, &queue_idx, "queue_id"),
        )
    });
    let mut queue_out = vec![queue_header];
    queue_out.extend(queue_body);
    fs::write(path, rows_to_csv(&queue_out))
        .map_err(|e| format!("failed to write {}: {e}", path.display()))?;

    Ok(payload().queue_entries.len())
}

fn check_sources(sources: &[SourceRecord]) -> Result<(), String> {
    let source_ids = sources
        .iter()
        .map(|source| source.id.clone())
        .collect::<HashSet<_>>();
    let mut missing = Vec::<String>::new();
    for entry in &payload().source_extract_entries {
        let source_id = entry.get("source_id").cloned().unwrap_or_default();
        if !source_id.is_empty() && !source_ids.contains(&source_id) {
            missing.push(source_id);
        }
    }
    if missing.is_empty() {
        Ok(())
    } else {
        Err(format!(
            "Missing source IDs in registry: {}",
            missing.join(", ")
        ))
    }
}

pub fn run(
    dataset: &Dataset,
    sources: &[SourceRecord],
    root_dir: &str,
    date_opt: Option<&str>,
) -> Result<String, String> {
    let date = date_opt.unwrap_or(DEFAULT_DATE);
    check_sources(sources)?;

    let root = Path::new(root_dir);
    let ledger_dir = root.join("docs").join("research-program").join("ledgers");
    let relationship_path = ledger_dir.join("relationship-evidence-ledger.csv");
    let source_extract_path = ledger_dir.join("source-extract-log.csv");
    let source_queue_path = ledger_dir.join("source-expansion-queue.csv");

    let claim_count = apply_relationship_updates(dataset, &relationship_path, date)?;
    let extract_count = upsert_source_extract_log(&source_extract_path, date)?;
    let queue_count = upsert_source_queue(&source_queue_path, date)?;

    Ok(format!(
        "Phase 1 locator batch A complete:\n- relationship claims updated: {claim_count}\n- source extract entries upserted: {extract_count}\n- source queue entries upserted: {queue_count}\n"
    ))
}
